promql/engine.go

```diff
@@ -1742,9 +1742,8 @@ func (ev *evaluator) eval(ctx context.Context, expr parser.Expr) (parser.Value,
 			ev.samplesStats.UpdatePeak(ev.currentSamples)
 
 			if e.Func.Name == "rate" || e.Func.Name == "increase" {
-				samples := inMatrix[0]
-				metricName := samples.Metric.Get(labels.MetricName)
-				if metricName != "" && len(samples.Floats) > 0 &&
+				metricName := inMatrix[0].Metric.Get(labels.MetricName)
+				if metricName != "" && len(ss.Floats) > 0 &&
```

Comment:
Can you explain in technical terms why it's right to check `len(ss.Floats)` and not `len(samples.Floats)`?

Reply:
`samples.Floats` could contain only 1 point which is not enough to output a rate (requires at least two points in the range vector). `ss.Floats` is the output, so it has determined that there were enough points selected to create a rate.

Of the (corrected) tests below, without this change the following fail:

- TestRateAnnotations/no_annotations_when_selecting_one_sample
- TestRateAnnotations/no_annotations_when_no_samples_due_to_mixed_data_types

So if your rate range vector selects no samples, no annotation is emitted.
If you selected two native histograms, no annotation is emitted.
If you selected 1 float and 1 native histogram, an annotation *was* emitted (and not longer is with this change).

The inconsistency is that either an annotation is emitted regardless if the selected metric has any output (eg even if `rate(non_existent_metric)` was called) or an annotation is emitted only if rate managed to return values. From the existing code I believe the latter to be the desired effect, otherwise the metric name could be checked at group creation time.

Reply:
> ss.Floats is the output, so it has determined that there were enough points selected to create a rate.

Ah, after debugging I understand what you mean. `ss.Floats` tells us whether `rate` actually could compute any samples.

Reply:
Your change makes sense to me!